### PR TITLE
Add tests section

### PR DESCRIPTION
**Pull request by justin-stephenson**

### CODE REVIEW

File: src/contrib/tests/unit-tests.rst [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+=============
+Unit tests
+=============
+
+Unit tests typically run a function or a tevent request without running the full deamon. The unit tests in SSSD are developed using either the `check <https://libcheck.github.io/check>`_ library or the `cmocka <https://cmocka.org>`_ library. Cmocka should be used for any new unit tests added to SSSD.
+
+The unit tests are fast to execute and in general this is where corner cases are typically easiest to test as you can provide false or unexpected input to the code under test. Unit tests are also often used to test a library's API.
+
+An important part of many tests using cmocka is wrapping a function provided by an external library using the `ld` linker's `--wrap` feature. You can learn more about cmocka and this feature in a `lwn.net article the cmocka developers contributed <https://lwn.net/Articles/558106/>`_. In the SSSD source tree, the unit tests reside under `src/tests/*.c` (check-based tests) and `src/tests/cmocka/*.c`.
+
```
Comment (pbrezina) [inline src/contrib/tests/unit-tests.rst:12]: Add files and code into double backquotes:
```
``code``
``` 

This will make it an inline code.

File: src/contrib/tests/unit-tests.rst [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+Examples of what can be tested by unit tests
+--------------------------------------------
+
+A typical use-case is the sysdb API tests at e.g. `src/tests/sysdb-tests.c`.
+
+A less typical use-case is testing of the NSS or PAM responders in isolation. The NSS responder test is located at `src/tests/cmocka/test_nss_srv.c`. Normally, the NSS responder would require a client such as getent to talk to it through the nss_sss module and would send requests to and receive replies from a back end. In unit tests, the NSS client's input is simulatd by calling the `sss_cmd_execute` directly, but with mocked input (see e.g. `mock_input_user_or_group`. The test even fakes communication to the Data Provider by mocking the `sss_dp_get_account_send` and `sss_dp_get_account_recv` request that normally talks to the Data Provider over D-Bus (see e.g. the `test_nss_getpwnam_search` test).
```
Comment (pbrezina) [inline src/contrib/tests/unit-tests.rst:25]: End parentheses is missing after mock_input_user_or_group.

File: src/contrib/tests/integration-tests.rst [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+The advantage over the unit tests is obvious, the full deamon is ran and you can talk to the SSSD using the same interfaces as a user would do in production, e.g. resolve a user with `getpwnam`. Because the tests are ran on the same machine as the developer works on, is is much faster to compile a new SSSD version for the tests to run and so the develop-test-fix cycle is generally quite fast. The integration tests also offer a simple way to add a "breakpoint" to the tests and connect to the tests using `screen(1)`. Finally, since the tests run on the same machine, they can trivially run on any OS release or any distribution with little to no changes, even in build systems that typically have no network connectivity as part of the SSSD build.
```
Comment (pbrezina) [inline src/contrib/tests/integration-tests.rst:9]: is is -> it is

File: src/contrib/tests/integration-tests.rst [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    $  screen -r sssd_cwrap_session
+
+From within the screen session, you can attach `gdb` to the SSSD processes, call `getent` to resolve users or groups `ldbsearch` the cache etc. To finish the debugging session, simply exit all the terminals in the tabs.
+
+Examples
+--------
+
+The tests themselves are located under `src/tests/intg`. Each file corresponds to one "test area", like testing the LDAP provider or testing the KCM responder.
+
+To see an example of adding test cases to existing tests, see commit `76ce965fc3abfdcf3a4a9518e57545ea060033d6` or for an example of adding a whole new test, including faking the client library (which should also illustrate the limits of the cwrap testing), see commit `5d838e13351d3062346ca449e00845750b9447da` and the two preceding it.
```
Comment (pbrezina) [inline src/contrib/tests/integration-tests.rst:112]: Link commit to github.

File: src/contrib/tests/multihost-tests.rst [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+The disadvantage of the tests is that setting up the environment can be complex and the development loop (the time between modifying test, modifying the SSSD sources, deploying them to the test environment and running the tests) is much longer than with the cwrap based tests.
+
+Running multihost tests
+-----------------------
+
+First, the infrastructure does not yet concern itself with provisioning at all. You need to set up a VM to run the tests on yourself. As an example, we'll be running the tests on a Fedora 28 machine which we will create using the `vagrant <https://www.vagrantup.com>`_ tool. To make things at least a little more palatable, the tests are expected to clean up after themselves, so it's possible to reuse the same provisioned VM for multiple test runs.
```
Comment (pbrezina) [inline src/contrib/tests/multihost-tests.rst:14]: Can we replace it with an example of running multihost test on sssd-test-suite?

File: src/contrib/running-tests.rst [hunk 6]
```diff
@@ -1,12 +1,20 @@
...
+Each test is different. Sometimes, you want to test the whole system end-to-end, sometimes the test should exercise some corner case for which special input and environment must be simulated. This section should give you a better idea what kind of tests already exist in SSSD so that you can choose where to add a new test and also provides a general overview.
 
-Etiam bibendum venenatis viverra. Aenean aliquet tortor diam, non fringilla enim luctus nec. Integer eros sem, sagittis vitae lectus ultrices, rhoncus sagittis ante. Interdum et malesuada fames ac ante ipsum primis in faucibus. Maecenas nec faucibus odio. Duis a placerat velit. Ut ac congue metus. Nunc ac enim convallis, condimentum ipsum id, mattis nulla. Maecenas quis consectetur nisi. Ut mi leo, dictum a nisi ut, cursus condimentum dolor. Aenean molestie diam quis justo aliquet fermentum. Quisque tellus turpis, luctus ut tempor vel, tincidunt bibendum augue. Maecenas vestibulum id dui et suscipit.
+* :ref:`unit-tests`
+* :ref:`intg-tests`
+* :ref:`multihost-tests`
```
Comment (pbrezina) [inline src/contrib/running-tests.rst:13]: Replace this with:
```
.. toctree::
    :maxdepth: 0
    :titlesonly:

    tests/unit-tests
    tests/integration-tests
    tests/multihost-tests
```

Then you can remove `type-list` completely.

**Review by pbrezina — CHANGES_REQUESTED** [review]

**Comment by justin-stephenson** [comment]
Thank you for the review, all items should be fixed now.

File: src/contrib/tests/multihost-tests.rst (hunk 5 above)
Comment (pbrezina) [reply src/contrib/tests/multihost-tests.rst]: You don't need to setup the root password now with sssd-test-suite. Also it is not currently possible to run other tests then `basic` (well it is possible but not straighforward, I still need to work on that). So only basic tests should be run now and you need to specify the directory so other tests are not picked up (it is currently not possible to run pytest from the root multihost dir).

Check this for guide: https://github.com/SSSD/sssd/blob/master/src/tests/multihost/README.md
But perhaps keep including the virtual environment (with a meaningful name?)

**Comment by pbrezina** [comment]
Thank you, I still have some comments to the multihost tests but it looks good otherwise.

**Comment by justin-stephenson** [comment]
Please see the latest multihost changes

File: src/contrib/tests/multihost-tests.rst [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    dnf install python3-pip python3-virtualenv
+    virtualenv .mh-venv
+    source .mh-venv/bin/activate
+
+The file ``src/tests/multihost/basic/mhc.yaml`` is already configured properly if you are using the SSSD Test Suite, run the tests with:
+
+.. code-block:: bash
+
+    sudo pip3 install -r src/tests/multihost/requirements.txt
```
Comment (pbrezina) [inline src/contrib/tests/multihost-tests.rst:35]: Remove sudo because you are in python venv.

**Review by pbrezina — CHANGES_REQUESTED** [review]
One last change and we can merge it.

**Comment by justin-stephenson** [comment]
> One last change and we can merge it.

Updated, thanks.

**Comment by pbrezina** [comment]
Thank you. Ack.
